internal/manifest/manifest.go

```diff
@@ -0,0 +1,212 @@
+package manifest
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"slices"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+const Filename = "manifest.wall"
+const Schema = "1.0"
+const Mode fs.FileMode = 0644
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+// LocateManifestSlices finds the paths marked with "generate:manifest" and
+// returns a map from the manifest path to all the slices that declare it.
+func LocateManifestSlices(slices []*setup.Slice) (map[string][]*setup.Slice, error) {
+	manifestSlices := make(map[string][]*setup.Slice)
+	for _, s := range slices {
+		for path, info := range s.Contents {
+			if info.Generate == setup.GenerateManifest {
+				dir, err := setup.GetGeneratePath(path)
+				if err != nil {
+					return nil, fmt.Errorf("internal error: %s", err)
+				}
+				path = filepath.Join(dir, Filename)
+				manifestSlices[path] = append(manifestSlices[path], s)
+			}
+		}
+	}
+	return manifestSlices, nil
+}
+
+type Manifest struct {
+	Paths    []Path
+	Contents []Content
+	Packages []Package
+	Slices   []Slice
+}
+
+func Read(rootDir string, relPath string) (mfest *Manifest, err error) {
```

Comment:
nitpick: How about `manifest` instead of `mfest`?
```suggestion
func Read(rootDir string, relPath string) (manifest *Manifest, err error) {
```

Reply:
Changed